File: apps/desktop/wm/qtile/conf/my_modules/key_map.py
"""short cut keys"""
from libqtile.config import Click, Drag, Key, KeyChord
from libqtile.lazy import lazy

from my_modules.param import PARAM
from my_modules.groups import GROUP_PER_SCREEN
from my_modules.my_functions import *

from libqtile.log_utils import logger


keys = [
    Key([PARAM.mod, 'shift'], 'Return', lazy.spawn(PARAM.terminal), desc='Launch terminal'),
    # Operate window in workspace
    Key([PARAM.mod], 'h', lazy.layout.left(), desc='Move focus to left'),
    Key([PARAM.mod], 'l', lazy.layout.right(), desc='Move focus to right'),
    Key([PARAM.mod], 'j', lazy.layout.down(), desc='Move focus down'),
    Key([PARAM.mod], 'k', lazy.layout.up(), desc='Move focus up'),
    Key([PARAM.mod, 'shift'], 'h', lazy.layout.shuffle_left(), desc='Move window to the left'),
    Key([PARAM.mod, 'shift'], 'l', lazy.layout.shuffle_right(), desc='Move window to the right'),
    Key([PARAM.mod, 'shift'], 'j', lazy.layout.shuffle_down(), lazy.layout.section_down(), desc='Move window down'),
    Key([PARAM.mod, 'shift'], 'k', lazy.layout.shuffle_up(), lazy.layout.section_up(), desc='Move window up'),
    Key([PARAM.mod, 'control'], 'c', lazy.window.kill(), desc='Kill focused window'),
    # Operate floating window
    Key([PARAM.mod], "period", float_cycle(forward=True)),
    Key([PARAM.mod], "comma", float_cycle(forward=True, focus=True)),
    # Operate window between workspaces
    Key([PARAM.mod, 'control'], 'h', focus_previous_group(), keep_pinp(), desc='focus prev group'),
    Key([PARAM.mod, 'control'], 'l', focus_next_group(), keep_pinp(), desc='focus next group'),
    Key([PARAM.mod, 'control'], 'j', window_to_previous_group(), keep_pinp(), desc='window to prev group'),
    Key([PARAM.mod, 'control'], 'k', window_to_next_group(), keep_pinp(), desc='window to next group'),
    # Operate screen
    Key([PARAM.mod], 'n', focus_cycle_screen(),  desc='focus next screen'),
    Key([PARAM.mod], 'p', focus_cycle_screen(backward=True),  desc='focus prev screen'),
    Key([PARAM.mod, 'shift'], 'n', move_cycle_screen(), keep_pinp(), update_pinp_screen_idx()),
    Key([PARAM.mod, 'shift'], 'p', move_cycle_screen(backward=True), keep_pinp(), update_pinp_screen_idx()),
    Key([PARAM.mod, 'shift'], 'd', to_from_display_tablet(), keep_pinp(), update_pinp_screen_idx()),
    # Toggle Float and FullScreen
    Key([PARAM.mod], 'space', lazy.window.toggle_maximize()),
    Key([PARAM.mod, 'shift'], 'space', lazy.window.toggle_minimize()),
    Key([PARAM.mod], 'f', lazy.window.toggle_fullscreen()),
    Key([PARAM.mod, 'shift'], 'f', lazy.window.toggle_floating()),
    # Exit and Reload
    Key([PARAM.mod, 'control'], 'r', lazy.reload_config(), desc='Reload the config'),
    Key([PARAM.mod, 'control'], 'e', lazy.shutdown(), desc='Shutdown Qtile'),
    # Launch rofi
    Key([PARAM.mod], 'm', lazy.spawn('rofi -combi-modi window,drun -show combi'), desc='show rofi'),
    Key([PARAM.mod, 'shift'], 'm', lazy.spawn('rofi -show run'), desc='run rofi as shell mode'),
    Key([PARAM.mod, 'control'], 'm', lazy.spawn('rofi -show mower-menu -modi mower-menu:rofi-mower-menu'), desc='show mower-menu'),
    # Lock screen
    Key([PARAM.mod, 'control'], 'b', lazy.spawn('i3lock -n -t -i {}'.format(PARAM.screen_saver)), desc='lock PC'),
    # Screenshot
    Key([], 'Print', lazy.spawn('flameshot full -p {}'.format(str(PARAM.capture_path)))),
    Key([PARAM.mod], 'Print', capture_screen(is_clipboard=False)),
    Key([PARAM.mod, 'shift'], 'Print', capture_screen(is_clipboard=True)),
    # Toggle copyq and calculator
    Key([PARAM.mod], 's', lazy.spawn('copyq toggle')),
    Key([PARAM.mod, 'shift'], 's', lazy.spawn('rofi -show calc -modi calc -no-show-match -no-sort')),
    # Function keys
    Key([], 'XF86AudioRaiseVolume', lazy.spawn('pactl set-sink-volume @DEFAULT_SINK@ +5%')),
    Key([], 'XF86AudioLowerVolume', lazy.spawn('pactl set-sink-volume @DEFAULT_SINK@ -5%')),
    Key([], 'XF86AudioMute', lazy.spawn('pactl set-sink-mute @DEFAULT_SINK@ toggle')),
    Key([], 'XF86AudioMicMute', lazy.spawn('pactl set-source-mute @DEFAULT_SOURCE@ toggle')),
    Key([], 'XF86AudioPlay', lazy.spawn('playerctl play-pause')),
    Key([], 'XF86AudioPause', lazy.spawn('playerctl play-pause')),
    Key([], 'XF86AudioNext', lazy.spawn('playerctl next')),
    Key([], 'XF86AudioPrev', lazy.spawn('playerctl previous')),
    Key([], 'XF86MonBrightnessUp', lazy.spawn('light -A 5')),
    Key([], 'XF86MonBrightnessDown', lazy.spawn('light -U 5')),
    Key([], 'XF86KbdBrightnessUp', lazy.spawn('light -Ars {} 1'.format('sysfs/leds/asus::kbd_backlight'))),
    Key([], 'XF86KbdBrightnessDown', lazy.spawn('light -Urs {} 1'.format('sysfs/leds/asus::kbd_backlight'))),
    Key([], 'XF86Launch1', lazy.spawn('sh -c "/home/sumi/bin/swich_gpu_mode"')),
    Key([], 'XF86Launch4', lazy.spawn('asusctl profile -n')),
    # Attach Screen
    KeyChord([PARAM.mod], 'a', [
        Key([], 'k', attach_screen('above')),
        Key([], 'j', attach_screen('below')),
        Key([], 'l', attach_screen('right-of')),
        Key([], 'h', attach_screen('left-of')),
        Key([], 'd', attach_screen('delete')),
        Key([], 'space', lazy.ungrab_chord()),
        ],
        name='attach'
    ),
    # PinP operationes
    KeyChord([PARAM.mod], 'v', [
        Key([], 'k', move_pinp('up')),
        Key([], 'j', move_pinp('down')),
        Key([], 'l', move_pinp('right')),
        Key([], 'h', move_pinp('left')),
        Key([], 'space', lazy.ungrab_chord()),
        ],
        name='pinp'
    )
]

for i in range(GROUP_PER_SCREEN):
    keys.extend([
        Key([PARAM.mod], str(i+1), focus_n_screen_group(i), keep_pinp,
            desc='Switch to group idx {}'.format(i+1)),

        Key([PARAM.mod, 'shift'], str(i+1), move_n_screen_group(i), keep_pinp,
            desc='Switch to & move focused window to group idx {}'.format(i+1)),
    ])

# Drag floating layouts.
mouse = [
    Drag([PARAM.mod], 'Button1', lazy.window.set_position_floating(),
         start=lazy.window.get_position()),
    Drag([PARAM.mod, 'shift'], 'Button1', lazy.window.set_size_floating(),
         start=lazy.window.get_size()),
]

File: apps/user/desktop/env/xmonad/conf/MyModules/Keymap.hs
import Data.default

import qualified XMonad.Actions.CycleWS as CWS
import qualified XMonad.Actions.DynamicWorkspaceOrder as DWSO
import qualified XMonad.Actions.PhysicalScreens as PS
import qualified XMonad.Actions.WindowNavigation as WN
import qualified XMonad.Operations as O
import qualified XMonad.StackSet as W --hiding (focusMaster, workspaces)
import qualified XMonad.Layout.MultiToggle.Instances as MTI
import qualified XMonad.Layout.ToggleLayouts as TL
import qualified XMonad.Terminal as T
import XMonad (mod4Mask, shiftMask, controlMask)
import XMonad (
    xK_Tab,
    xK_Return,
    xK_space,
    xK_h,
    xK_j,
    xK_k,
    xK_l,
    xK_n,
    xK_p,
    xK_q,
    xK_f,
    xK_c,
    xK_m,
    xK_a,
    xK_v,
    xK_r,
    xK_e,
    xK_comma,
    xK_period,
    xK_Print
)
import qualified WorkspacePerMonitor as WPM
import qualified Params as P


toggleFloat x w = O.windows (\s -> if M.member w (W.floating s)
                            then W.sink w s
                            else W.float w (W.RationalRect (1/3) (1/4) (1/2) (4/5) s))


modm = mod4Mask  --Super Key

keymap = [
    -- Operate window in workspace
      ((modm,                    xK_h), O.sendMessage $ WN.Go WN.L)
    , ((modm,                    xK_j), O.sendMessage $ WN.Go WN.D)
    , ((modm,                    xK_k), O.sendMessage $ WN.Go WN.U)
    , ((modm,                    xK_l), O.sendMessage $ WN.Go WN.R)
    , ((modm .|. shiftMask,      xK_h), O.sendMessage $ WN.Swap WN.L)
    , ((modm .|. shiftMask,      xK_j), O.sendMessage $ WN.Swap WN.D)
    , ((modm .|. shiftMask,      xK_k), O.sendMessage $ WN.Swap WN.U)
    , ((modm .|. shiftMask,      xK_l), O.sendMessage $ WN.Swap WN.R)
    , ((modm .|. controlMask   xK_c), O.kill)
     -- Operate window between workspaces
    , ((modm .|. controlMask,    xK_h), DWSO.moveTo CWS.Prev WPM.spacesOnCurrentScreen)
    , ((modm .|. controlMask,    xK_l), DWSO.moveTo CWS.Next WPM.spacesOnCurrentScreen)
    , ((modm .|. controlMask,    xK_j), DWSO.shiftTo CWS.Prev WPM.spacesOnCurrentScreen
                                        >> DWSO.moveTo CWS.Prev WPM.spacesOnCurrentScreen)
    , ((modm .|. controlMask,    xK_k), DWSO.shiftTo CWS.Next WPM.spacesOnCurrentScreen
    -- Operate screen
    , ((modm,                  xK_n), PS.onNextNeighbour def W.view)
    , ((modm .|. shiftMask,    xK_n), PS.onNextNeighbour def W.shift)
    , ((modm,                  xK_p), PS.onPrevNeighbour def W.view)
    , ((modm .|. shiftMask,    xK_p), PS.onPrevNeighbour def W.shift)
    -- Toggle float and FullScreen
    , ((modm,                  xK_space), O.sendMessage $ TL.Toggle MTI.NBFULL)
    , ((modm,                  xK_f), O.sendMessage $ TL.Toggle MTI.NBFULL)
    , ((modm .|. shiftMask,    xK_f), O.withFocused $ ToggleFloat)
    -- Exit and Reload
    , ((modm .|. controlMask,  xK_e), io exitSuccess)
    , ((modm .|. controlMask,  xK_r), spawn "if type xmonad; then xmonad --recompile && xmonad --restart; else xmessage xmonad not in \\$PATH: \"$PATH\"; fi")
    -- Launch rofi
    , ((modm,                  xK_m), spawn "rofi -combi-modi window,drun -show combi")
    , ((modm .|. shiftMask,    xK_m), spawn "rofi -show run")
    , ((modm .|. controlMask,  xK_m), spawn "rofi -show mower-menu -modi mower-menu:rofi-mower-menu")
    -- Lock screen
    , ((modm .|. controlMask,  xK_b), spawn "i3lock -n -t -i " ++ P.screenSaver)
    -- Screenshot
    , ((modm,                  xK_Print), spawn "flameshot full -p " ++ P.capturePath)
    , ((modm .|. shiftMask,    xK_Print), spawn "flameshot screen -c")
    , ((modm .|. controlMask,  xK_Print), spawn "flameshot screen -p " ++ P.capturePath)
    -- Show copyq and calculator
    , ((modm,                  xK_s), spawn "copyq toggle")
    , ((modm .|. shiftMask,    xK_s), spawn "rofi -show calc -modi calc -no-show-match -no-sort")
]

functionMap = [
      ("<XF86AudioRaiseVolume>", spawn "pactl set-sink-volume @DEFAULT_SINK@ +5%")
    , ("<XF86AudioLowerVolume>", spawn "pactl set-sink-volume @DEFAULT_SINK@ -5%")
    , ("<XF86AudioMute>", spawn "pactl set-sink-mute @DEFAULT_SINK@ toggle")
    , ("<XF86AudioMicMute>", spawn "pactl set-source-mute @DEFAULT_SOURCE@ toggle")
    , ("<XF86AudioPlay>", spawn "playerctl play-play")
    , ("<XF86AudioPause>", spawn "playerctl play-pause")
    , ("<XF86AudioNext>", spawn "playerctl next")
    , ("<XF86AudioPrev>", spawn "playerctl previous")
    , ("<XF86MonBrightnessUp>", spawn "light -A 5")
    , ("<XF86MonBrightnessDown>", spawn "light -U 5")
    , ("<XF86KbdBrightnessUp>", spawn "light -Ars 'sysfs/leds/asus::kbd_backlight' 1")
    , ("<XF86KbdBrightnessDown>", spawn "light -Urs 'sysfs/leds/asus::kbd_backlight' 1")
    , ("<XF86Launch1>", spawn "sh -c "/home/sumi/bin/swich_gpu_mode")
    , ("<XF86Launch4>", spawn "asusctl profile -n")
]
